### PR TITLE
Refactor reshare tests

### PR DESCRIPTION
**Pull request by akoshelev**
While working on #256, I noticed that reshare tests consistently fail concurrency test. It was caused by always using RecordId(0) in a loop that terminates if reshare successfully generates unique shares. When loop needs more than 1 iteration it fails on PRSS uniqueness check.

While it was possible to fix that test, I did not really like the probabilistic nature of it and decided to create 2 tests instead. One validates that reshare protocol actually communicates and another one that it is correct.

It was also a good opportunity to use new test API implemented by @martinthomson in #261. It should be used everywhere.

One small change I made to the `TestWorld` to improve ergonomics changes `make_contexts` function to narrow generated contexts to some unique (within this world) step, so the following syntax is made possible

```rust
  for _ in 0..10 {
     world.semi_honest(..., |ctx, v| { ... })
  }

### CODE REVIEW

File: src/protocol/sort/reshare.rs [hunk 1]
```diff
@@ -81,52 +81,77 @@ impl<F: Field> Reshare<F> {
...
+        for &target in Role::all() {
+            let secret = thread_rng().gen::<Fp32BitPrime>();
+            let shares = world
+                .semi_honest(secret, |ctx, share| async move {
+                    let record_id = RecordId::from(0);
+
+                    // run reshare protocol for all helpers except the one that does not know the input
+                    if ctx.role() == target {
+                        // test follows the reshare protocol
+                        ctx.prss().generate_fields(record_id).into()
```
Comment (martinthomson) [inline src/protocol/sort/reshare.rs:113]: Won't the effect be the same if you run reshare for this target?  Or, are you trying to test that it doesn't need to participate?

File: src/protocol/sort/reshare.rs [hunk 2]
```diff
@@ -81,52 +81,77 @@ impl<F: Field> Reshare<F> {
...
+                    } else {
+                        Reshare::new(share.clone())
+                            .execute(&ctx, record_id, target)
```
Comment (martinthomson) [inline src/protocol/sort/reshare.rs:116]: This still takes a reference to the context, when it shouldn't.

Comment (akoshelev) [reply src/protocol/sort/reshare.rs:116]: let me merge #257 and rebase this on top of that

File: src/protocol/sort/reshare.rs (hunk 1 above)
Comment (akoshelev) [reply src/protocol/sort/reshare.rs:113]: I tried to ensure a naive `Reshare` implementation (that is - take the input and return it as output) won't pass this test. I think that was the reason for `for` loop inside the test before.

**Review by martinthomson — APPROVED** [review]

**Pull request merged**